python/ray/serve/controller.py

```diff
@@ -76,6 +83,386 @@ class Config:
         arbitrary_types_allowed = True
 
 
+@dataclass
+class ConfigurationStore:
+    backends: Dict[BackendTag, BackendInfo]
+    traffic_policies: Dict[EndpointTag, TrafficPolicy]
+    routes: Dict[BackendTag, Tuple[EndpointTag, Any]]
+
+    def __init__(self):
+        self.backends = dict()
+        self.traffic_policies = dict()
+        self.routes = dict()
+
+    def get_backend_configs(self) -> Dict[BackendTag, BackendConfig]:
+        result = {}
+        for tag, info in self.backends.items():
+            config = info.backend_config
+            result[tag] = config
+        return result
+
+    def get_backend(self, backend_tag: BackendTag) -> Optional[BackendInfo]:
+        return self.backends.get(backend_tag, None)
+
+    def add_backend(self, backend_tag: BackendTag,
+                    backend_info: BackendInfo) -> None:
+        self.backends[backend_tag] = backend_info
+
+
+@dataclass
+class ActorStateReconciler:
+    routers_cache: Dict[NodeId, ActorHandle]
+    replicas: Dict[BackendTag, List[ReplicaTag]]
+    replicas_to_start: Dict[BackendTag, List[ReplicaTag]]
+    replicas_to_stop: Dict[BackendTag, List[ReplicaTag]]
+    backends_to_remove: List[BackendTag]
+    endpoints_to_remove: List[EndpointTag]
+    # TODO(edoakes): consider removing this and just using the names.
+    workers: Dict[BackendTag, Dict[ReplicaConfig, ActorHandle]]
+    controller_name: str
+    detached: bool
+
+    def __init__(self, controller_name: str, detached: bool):
+        self.routers_cache = dict()
+        self.replicas = defaultdict(list)
+        self.replicas_to_start = defaultdict(list)
+        self.replicas_to_stop = defaultdict(list)
+        self.backends_to_remove = list()
+        self.endpoints_to_remove = list()
+        self.workers = defaultdict(dict)
+        controller_name = controller_name
+        detached = detached
+
+    def router_handles(self) -> List[ActorHandle]:
+        return self.routers_cache.values()
```

Comment:
this won't pass typecheck, dict.values() returns some set-like object. wrap it to list(self.router_cache.values()) for list